### PR TITLE
add support for nesterov momentum

### PR DESCRIPTION
**Pull request by suiyuan2009**
#594
- I use an indirect way. After every minimal step, var will be var_true - lr \* momentum \* accum, so we need to recover to var_true before compute new var.
- I find it's difficult to implement  nesterov momentum in a direct way due to the code structure of minimal step.
- The first commit only include dense nesterov momentum, I'll add sparse version and docs.
- The code will not pass test currently becase of accuracy problem of tf.half.
- **Anyone has a better suggestion?**

### CODE REVIEW

**Comment by georgedahl** [comment]
Can you add a derivation/explanation on how the code correctly implements the Nesterov-style momentum updates? On the face of it, it isn't clear how the correct gradient gets computed so explaining the reparameterizaiton you use and proving it is equivalent to the obvious implementation would be helpful.

**Comment by suiyuan2009** [comment]
@georgedahl 
the obvious implementation of nesterov momentum is as follow
```
1. var' = var - lr_ * momentum_ * accum
2. grad = some_function(var')
3. accum = momentum_ * accum + grad
4. var = var - lr_ * accum
```
In tensorflow, a minimal step contains two parts, one is compute_gradients, another is apply_gradients. In my implenmentation, after apply_gradients, the var stored is actually var - lr_ * momentum_ * accum, so the value of var' will pass to compute_gradients. In the next apply_gradients step, we should first recover var by var = var + lr_ * momentum_ * accum. I combined formulas to simplify codes.

**Comment by suiyuan2009** [comment]
@georgedahl I added sparse version and docs, removed tf.half in test.

**Comment by georgedahl** [comment]
I am going to ICML and I won't get to look at this until I return.

**Comment by vrv** [comment]
ping for @georgedahl

**Comment by georgedahl** [comment]
Sorry to take so long to look at this. I believe this logic is correct and is fine to merge. I defer to @vrv for any other cleanup.

File: tensorflow/core/ops/training_ops.cc [hunk 1]
```diff
@@ -356,6 +358,9 @@ out: Same as "var".
 use_locking: If `True`, updating of the var and accum tensors will be protected
   by a lock; otherwise the behavior is undefined, but may exhibit less
   contention.
+use_nesterov: If `True`, the var pass to compute grad will be
```
Comment (vrv) [inline tensorflow/core/ops/training_ops.cc]: the tensor passed to compute

File: tensorflow/core/kernels/training_ops.cc [hunk 2]
```diff
@@ -176,9 +175,13 @@ struct ApplyMomentum<CPUDevice, T> {
                   typename TTypes<T>::Flat accum,
                   typename TTypes<T>::ConstScalar lr,
                   typename TTypes<T>::ConstFlat grad,
-                  typename TTypes<T>::ConstScalar momentum) {
+                  typename TTypes<T>::ConstScalar momentum, bool use_nesterov) {
     accum.device(d) = accum * momentum() + grad;
-    var.device(d) -= accum * lr();
+    if (use_nesterov == true) {
```
Comment (vrv) [inline tensorflow/core/kernels/training_ops.cc]: just if (use_nesterov) {  should be fine

File: tensorflow/core/kernels/training_ops.cc [hunk 3]
```diff
@@ -1672,7 +1678,12 @@ class SparseApplyMomentumOp : public OpKernel {
         auto g = grad_flat.template chip<0>(i);
         auto v = var_flat.template chip<0>(index);
         a = a * a.constant(momentum_scalar) + g;
-        v -= a.constant(lr_scalar) * a;
+        if (use_nesterov_ == true) {
```
Comment (vrv) [inline tensorflow/core/kernels/training_ops.cc]: (same here)

**Comment by suiyuan2009** [comment]
@vrv cleanups done

File: tensorflow/core/ops/training_ops.cc (hunk 1 above)
Comment (vrv) [inline tensorflow/core/ops/training_ops.cc]: (i don't think this was fixed?)

Comment (suiyuan2009) [inline tensorflow/core/ops/training_ops.cc]: fixed now

**Comment by suiyuan2009** [comment]
ping for @vrv

**Comment by vrv** [comment]
@tensorflow-jenkins test this please

**Comment by vrv** [comment]
Looks like the GPU implementation also needs these changes -- can you make the appropriate changes to the GPU version of the functor?

**Comment by suiyuan2009** [comment]
I'll make change to the functor in `tensorflow/core/kernels/training_ops_gpu.cu.cc`, may be I need to find a machine with nvidia gpu since my laptop doesn't support it.

**Comment by suiyuan2009** [comment]
gpu version functor is ok now

**Comment by suiyuan2009** [comment]
ping for @vrv

File: tensorflow/core/ops/training_ops.cc [hunk 4]
```diff
@@ -341,8 +341,10 @@ REGISTER_OP("ApplyMomentum")
     .Output("out: Ref(T)")
     .Attr("T: numbertype")
     .Attr("use_locking: bool = false")
+    .Attr("use_nesterov: bool = false")
     .Doc(R"doc(
-Update '*var' according to the momentum scheme.
+Update '*var' according to the momentum scheme. Set use_nesterov = True if you 
+want to use nesterov momentum.
```
Comment (vrv) [inline tensorflow/core/ops/training_ops.cc]: capitalize Nesterov

File: tensorflow/core/ops/training_ops.cc [hunk 5]
```diff
@@ -369,8 +374,10 @@ REGISTER_OP("SparseApplyMomentum")
     .Attr("T: numbertype")
     .Attr("Tindices: {int32, int64}")
     .Attr("use_locking: bool = false")
+    .Attr("use_nesterov: bool = false")
     .Doc(R"doc(
 Update relevant entries in '*var' and '*accum' according to the momentum scheme.
+Set use_nesterov = True if you want to use nesterov momentum.
```
Comment (vrv) [inline tensorflow/core/ops/training_ops.cc]: same here

**Comment by vrv** [comment]
@tensorflow-jenkins test this please

**Comment by vrv** [comment]
@tensorflow-jenkins test this please

**Comment by avostryakov** [comment]
@suiyuan2009 It's pity but Nesterov Momentum doesn't work with embedding matrix on GPU. The error: "Cannot assign a device to node 'gradients/networks/transpose_1_grad/transpose/x': Could not satisfy explicit device specification '/device:GPU:0' because no supported kernel for GPU devices is available."

Maybe you can fix it?

**Comment by suiyuan2009** [comment]
@avostryakov have you seen #2285 and #2292? if not, try adding "allow_soft_placement=True".

**Comment by avostryakov** [comment]
@suiyuan2009 It's exactly what I meant. I can use allow_soft_placement=True but things are become extremely slow during training because embedding matrix will be stored on CPU device

**Comment by avostryakov** [comment]
I saw that you worked with RMSProb optimizer on the same problem in #2564. It's why I thought that you can do the same here.

**Comment by suiyuan2009** [comment]
@avostryakov the RMSProb optimizer that I write is a sparse version, now it only supports CPU device.
@vrv can you give some advice to @avostryakov ?
I'd like to take a deep look at this GPU bug.

**Comment by suiyuan2009** [comment]
@avostryakov you can try distributed tensorflow, but take care of memory management of your GPUs.

**Comment by avostryakov** [comment]
@suiyuan2009 What is the point to use distributed tensorflow? Do you mean on several machines with GPUs?

**Comment by vrv** [comment]
Yeah, I don't think any of the sparse optimizer training ops are implemented on GPU.  It would be awesome if someone wanted to do this, though it probably requires either cuda or Eigen programming knowledge to implement.

**Comment by avostryakov** [comment]
@vrv I don't have skills in this field but yeah, it could be awesome! By the way, AdamOptimizer can train sparse tensors on GPU, isn't it?

**Comment by vrv** [comment]
I don't see an implementation of SparseApplyAdam on GPU, so unfortunately not.  Until such operations are implemented, it's best to keep the variables on CPU and the computation on GPU.

**Comment by suiyuan2009** [comment]
@avostryakov you can use distributed tensorflow on single machine,[distributed tensorflow](https://www.tensorflow.org/versions/r0.9/how_tos/distributed/index.html#distributed-tensorflow), I guess you run your code on one machine with many GPUs.

**Comment by avostryakov** [comment]
@vrv Thank you for the reply!
